File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/BroadCaster.java
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.Command;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;

import java.util.Map;
import java.util.*;

public class BroadCaster extends ICommand {

    public BroadCaster(String name) {
        super(name);
    }

    @Override
    public void onPlayerCommand(Player sp, Command cmd, String label, String[] args) {
        if (label.equalsIgnoreCase("broadcaster") || label.equalsIgnoreCase("yt")) {
            if (TosoGameAPI.isAdmin(sp)) {
                if (args.length != 0) {
                    if (args[0].equalsIgnoreCase("add")) {
                        if (args.length != 1) {
                            Player target = Bukkit.getPlayerExact(args[1]);
                            if (target == null) {
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                                return;
                            } else {
                                Player p = Bukkit.getPlayerExact(args[1]);
                                if (!TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).setBroadCaster(true);

                                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者に追加しました。");
                                    return;
                                }
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者になっています。");
                                return;
                            }
                        }
                        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.ARGS_PLAYER);
                        return;
                    } else if (args[0].equalsIgnoreCase("remove")) {
                        if (args.length != 1) {
                            Player target = Bukkit.getPlayerExact(args[1]);
                            if (target == null) {
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                                return;
                            } else {
                                Player p = Bukkit.getPlayerExact(args[1]);
                                if (TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).setBroadCaster(false);

                                    sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者から削除しました。");
                                    return;
                                }
                                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者ではありません。");
                                return;
                            }
                        }
                        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.ARGS_PLAYER);
                        return;
                    } else if (args[0].equalsIgnoreCase("list")) {
                        StringBuffer stringBuffer = new StringBuffer();
                        for (Map.Entry<UUID, PlayerConfig> entry : PlayerManager.getConfigs())
                            if (entry.getValue().getBroadCaster())
                                stringBuffer.append(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + Bukkit.getOfflinePlayer(entry.getKey()).getName() + "\n");

                        sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.WARNING) + "配信者リスト:\n" +
                                stringBuffer.toString().trim());
                        return;
                    }
                }
                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster add <プレイヤー名>\" - 配信者を追加\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster remove <プレイヤー名>\" - 配信者を削除\n" +
                        MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster list\" - 配信者リストを表示");
                return;
            }
            MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
            return;
        } else if (label.equalsIgnoreCase("lhide")) {
            if (TosoGameAPI.isBroadCaster(sp)) {
                TosoGameAPI.addHidePlayer(sp);

                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "あなたを非表示にしました。");
                return;
            }
            MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
            return;
        } else if (label.equalsIgnoreCase("lshow")) {
            if (TosoGameAPI.isBroadCaster(sp)) {
                TosoGameAPI.removeHidePlayer(sp);

                sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + "あなたを表示しました。");
                return;
            }
            MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
            return;
        }
    }

    @Override
    public void onBlockCommand(BlockCommandSender bs, Command cmd, String label, String[] args) {
        if (label.equalsIgnoreCase("broadcaster") || label.equalsIgnoreCase("yt")) {
            if (args.length != 0) {
                if (args[0].equalsIgnoreCase("add")) {
                    if (args.length != 1) {
                        Player target = Bukkit.getPlayerExact(args[1]);
                        if (target == null) {
                            bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                            return;
                        } else {
                            Player p = Bukkit.getPlayerExact(args[1]);
                            if (!TosoGameAPI.isBroadCaster(p)) {
                                PlayerManager.loadConfig(p).setBroadCaster(true);

                                bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者に追加しました。");
                                return;
                            }
                            bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者になっています。");
                            return;
                        }
                    }
                    MainAPI.sendMessage(bs, MainAPI.ErrorMessage.ARGS_PLAYER);
                    return;
                } else if (args[0].equalsIgnoreCase("remove")) {
                    if (args.length != 1) {
                        Player target = Bukkit.getPlayerExact(args[1]);
                        if (target == null) {
                            bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                            return;
                        } else {
                            Player p = Bukkit.getPlayerExact(args[1]);
                            if (TosoGameAPI.isBroadCaster(p)) {
                                PlayerManager.loadConfig(p).setBroadCaster(false);

                                bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者から削除しました。");
                                return;
                            }
                            bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者ではありません。");
                            return;
                        }
                    }
                    MainAPI.sendMessage(bs, MainAPI.ErrorMessage.ARGS_PLAYER);
                    return;
                } else if (args[0].equalsIgnoreCase("list")) {
                    StringBuffer stringBuffer = new StringBuffer();
                    for (Map.Entry<UUID, PlayerConfig> entry : PlayerManager.getConfigs())
                        if (entry.getValue().getBroadCaster())
                            stringBuffer.append(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + Bukkit.getOfflinePlayer(entry.getKey()).getName() + "\n");

                    bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.WARNING) + "配信者リスト:\n" +
                            stringBuffer.toString().trim());
                    return;
                }
            }
            bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster add <プレイヤー名>\" - 配信者を追加\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster remove <プレイヤー名>\" - 配信者を削除\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster list\" - 配信者リストを表示");
            return;
        } else if (label.equalsIgnoreCase("lhide") || label.equalsIgnoreCase("lshow")) {
            MainAPI.sendMessage(bs, MainAPI.ErrorMessage.NOT_PLAYER);
            return;
        }
    }

    @Override
    public void onConsoleCommand(ConsoleCommandSender cs, Command cmd, String label, String[] args) {
        if (label.equalsIgnoreCase("broadcaster") || label.equalsIgnoreCase("yt")) {
            if (args.length != 0) {
                if (args[0].equalsIgnoreCase("add")) {
                    if (args.length != 1) {
                        Player target = Bukkit.getPlayerExact(args[1]);
                        if (target == null) {
                            cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                            return;
                        } else {
                            Player p = Bukkit.getPlayerExact(args[1]);
                            if (!TosoGameAPI.isBroadCaster(p)) {
                                PlayerManager.loadConfig(p).setBroadCaster(true);

                                cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者に追加しました。");
                                return;
                            }
                            cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者になっています。");
                            return;
                        }
                    }
                    MainAPI.sendMessage(cs, MainAPI.ErrorMessage.ARGS_PLAYER);
                    return;
                } else if (args[0].equalsIgnoreCase("remove")) {
                    if (args.length != 1) {
                        Player target = Bukkit.getPlayerExact(args[1]);
                        if (target == null) {
                            cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + ChatColor.RED + ChatColor.UNDERLINE + args[1] + ChatColor.RESET + ChatColor.GRAY + "はオフラインです。");
                            return;
                        } else {
                            Player p = Bukkit.getPlayerExact(args[1]);
                            if (TosoGameAPI.isBroadCaster(p)) {
                                PlayerManager.loadConfig(p).setBroadCaster(false);

                                cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.SUCCESS) + p.getName() + "を配信者から削除しました。");
                                return;
                            }
                            cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + p.getName() + "はすでに配信者ではありません。");
                            return;
                        }
                    }
                    MainAPI.sendMessage(cs, MainAPI.ErrorMessage.ARGS_PLAYER);
                    return;
                } else if (args[0].equalsIgnoreCase("list")) {
                    StringBuffer stringBuffer = new StringBuffer();
                    for (Map.Entry<UUID, PlayerConfig> entry : PlayerManager.getConfigs())
                        if (entry.getValue().getBroadCaster())
                            stringBuffer.append(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + Bukkit.getOfflinePlayer(entry.getKey()).getName() + "\n");

                    cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.WARNING) + "配信者リスト:\n" +
                            stringBuffer.toString().trim());
                    return;
                }
            }
            cs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "引数が不正です。\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "コマンドの使い方:\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster add <プレイヤー名>\" - 配信者を追加\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster remove <プレイヤー名>\" - 配信者を削除\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "\"/broadcaster list\" - 配信者リストを表示");
            return;
        } else if (label.equalsIgnoreCase("lhide") || label.equalsIgnoreCase("lshow")) {
            MainAPI.sendMessage(cs, MainAPI.ErrorMessage.NOT_PLAYER);
            return;
        }
    }

    @Override
    public List<String> onPlayerTabComplete(Player sp, Command cmd, String alias, String[] args) {
        if (!TosoGameAPI.isAdmin(sp) || args.length != 1) return null;
        return getTabList(args[0], new HashSet<>(Arrays.asList("add", "remove", "list")));
    }

    @Override
    public List<String> onBlockTabComplete(BlockCommandSender bs, Command cmd, String alias, String[] args) {
        if (args.length != 1) return null;
        return getTabList(args[0], new HashSet<>(Arrays.asList("add", "remove", "list")));
    }

    @Override
    public List<String> onConsoleTabComplete(ConsoleCommandSender cs, Command cmd, String alias, String[] args) {
        if (args.length != 1) return null;
        return getTabList(args[0], new HashSet<>(Arrays.asList("add", "remove", "list")));
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Broadcaster.kt
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.ErrorMessage
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.VisibilityManager
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand
import jp.aoichaan0513.A_TosoGame_Live.Utils.toString
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.Command
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player

class Broadcaster(name: String) : ICommand(name) {

    override fun onPlayerCommand(sp: Player, cmd: Command, label: String, args: Array<String>) {
        if (label.equals("broadcaster", true) || label.equals("yt", true)) {
            if (TosoGameAPI.isAdmin(sp)) {
                if (args.isNotEmpty()) {
                    when (args[0].toLowerCase()) {
                        "add" -> {
                            if (args.size != 1) {
                                val p = Bukkit.getPlayerExact(args[1])
                                if (p != null) {
                                    if (!TosoGameAPI.isBroadCaster(p)) {
                                        PlayerManager.loadConfig(p).broadCaster = true
                                        sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者に追加しました。")
                                        return
                                    }
                                    sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者になっています。")
                                    return
                                }
                                MainAPI.sendOfflineMessage(sp, args[1])
                                return
                            }
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_PLAYER)
                        }
                        "remove" -> {
                            if (args.size != 1) {
                                val p = Bukkit.getPlayerExact(args[1])
                                if (p != null) {
                                    if (TosoGameAPI.isBroadCaster(p)) {
                                        PlayerManager.loadConfig(p).broadCaster = false
                                        sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者から削除しました。")
                                        return
                                    }
                                    sp.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者ではありません。")
                                    return
                                }
                                MainAPI.sendOfflineMessage(sp, args[1])
                                return
                            }
                            MainAPI.sendMessage(sp, ErrorMessage.ARGS_PLAYER)
                        }
                        "list" -> {
                            sp.sendMessage("${MainAPI.getPrefix(PrefixType.WARNING)}配信者リスト: ${MainAPI.getOnlinePlayers(PlayerManager.configs.filter { it.value.broadCaster }.map { it.key }).toString({ "${ChatColor.YELLOW}${it.name}${ChatColor.GRAY}" }, "")}")
                        }
                    }
                    return
                }
                sp.sendMessage("""
                    ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                    ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster add <プレイヤー名>" - 配信者を追加
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster remove <プレイヤー名>" - 配信者を削除
                    ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster list" - 配信者リストを表示
                """.trimIndent())
                return
            }
            MainAPI.sendMessage(sp, ErrorMessage.PERMISSIONS_TEAM_ADMIN)
        } else if (label.equals("lhide", true)) {
            if (TosoGameAPI.isBroadCaster(sp)) {
                VisibilityManager.add(sp, VisibilityManager.VisibilityType.LIVE)
                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}あなたを非表示にしました。")
                return
            }
            MainAPI.sendMessage(sp, ErrorMessage.PERMISSIONS)
        } else if (label.equals("lshow", true)) {
            if (TosoGameAPI.isBroadCaster(sp)) {
                VisibilityManager.remove(sp, VisibilityManager.VisibilityType.LIVE)
                sp.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}あなたを表示しました。")
                return
            }
            MainAPI.sendMessage(sp, ErrorMessage.PERMISSIONS)
        }
    }

    override fun onBlockCommand(bs: BlockCommandSender, cmd: Command, label: String, args: Array<String>) {
        if (label.equals("broadcaster", true) || label.equals("yt", true)) {
            if (args.isNotEmpty()) {
                when (args[0].toLowerCase()) {
                    "add" -> {
                        if (args.size != 1) {
                            val p = Bukkit.getPlayerExact(args[1])
                            if (p != null) {
                                if (!TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).broadCaster = true
                                    bs.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者に追加しました。")
                                    return
                                }
                                bs.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者になっています。")
                                return
                            }
                            MainAPI.sendOfflineMessage(bs, args[1])
                            return
                        }
                        MainAPI.sendMessage(bs, ErrorMessage.ARGS_PLAYER)
                    }
                    "remove" -> {
                        if (args.size != 1) {
                            val p = Bukkit.getPlayerExact(args[1])
                            if (p != null) {
                                if (TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).broadCaster = false
                                    bs.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者から削除しました。")
                                    return
                                }
                                bs.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者ではありません。")
                                return
                            }
                            MainAPI.sendOfflineMessage(bs, args[1])
                            return
                        }
                        MainAPI.sendMessage(bs, ErrorMessage.ARGS_PLAYER)
                    }
                    "list" -> {
                        bs.sendMessage("${MainAPI.getPrefix(PrefixType.WARNING)}配信者リスト: ${MainAPI.getOnlinePlayers(PlayerManager.configs.filter { it.value.broadCaster }.map { it.key }).toString({ "${ChatColor.YELLOW}${it.name}${ChatColor.GRAY}" }, "")}")
                    }
                }
                return
            }
            bs.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster add <プレイヤー名>" - 配信者を追加
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster remove <プレイヤー名>" - 配信者を削除
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster list" - 配信者リストを表示
            """.trimIndent())
        } else if (label.equals("lhide", true) || label.equals("lshow", true)) {
            MainAPI.sendMessage(bs, ErrorMessage.NOT_PLAYER)
        }
    }

    override fun onConsoleCommand(cs: ConsoleCommandSender, cmd: Command, label: String, args: Array<String>) {
        if (label.equals("broadcaster", true) || label.equals("yt", true)) {
            if (args.isNotEmpty()) {
                when (args[0].toLowerCase()) {
                    "add" -> {
                        if (args.size != 1) {
                            val p = Bukkit.getPlayerExact(args[1])
                            if (p != null) {
                                if (!TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).broadCaster = true
                                    cs.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者に追加しました。")
                                    return
                                }
                                cs.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者になっています。")
                                return
                            }
                            MainAPI.sendOfflineMessage(cs, args[1])
                            return
                        }
                        MainAPI.sendMessage(cs, ErrorMessage.ARGS_PLAYER)
                    }
                    "remove" -> {
                        if (args.size != 1) {
                            val p = Bukkit.getPlayerExact(args[1])
                            if (p != null) {
                                if (TosoGameAPI.isBroadCaster(p)) {
                                    PlayerManager.loadConfig(p).broadCaster = false
                                    cs.sendMessage("${MainAPI.getPrefix(PrefixType.SUCCESS)}${p.name}を配信者から削除しました。")
                                    return
                                }
                                cs.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}${p.name}はすでに配信者ではありません。")
                                return
                            }
                            MainAPI.sendOfflineMessage(cs, args[1])
                            return
                        }
                        MainAPI.sendMessage(cs, ErrorMessage.ARGS_PLAYER)
                    }
                    "list" -> {
                        cs.sendMessage("${MainAPI.getPrefix(PrefixType.WARNING)}配信者リスト: ${MainAPI.getOnlinePlayers(PlayerManager.configs.filter { it.value.broadCaster }.map { it.key }).toString({ "${ChatColor.YELLOW}${it.name}${ChatColor.GRAY}" }, "")}")
                    }
                }
                return
            }
            cs.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.ERROR)}引数が不正です。
                ${MainAPI.getPrefix(PrefixType.ERROR)}コマンドの使い方:
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster add <プレイヤー名>" - 配信者を追加
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster remove <プレイヤー名>" - 配信者を削除
                ${MainAPI.getPrefix(PrefixType.ERROR)}"/broadcaster list" - 配信者リストを表示
            """.trimIndent())
        } else if (label.equals("lhide", true) || label.equals("lshow", true)) {
            MainAPI.sendMessage(cs, ErrorMessage.NOT_PLAYER)
        }
    }

    override fun onPlayerTabComplete(sp: Player, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (!TosoGameAPI.isAdmin(sp) || args.size != 1) return emptyList()
        return getTabList(args[0], "add", "remove", "list")
    }

    override fun onBlockTabComplete(bs: BlockCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (args.size != 1) return emptyList()
        return getTabList(args[0], "add", "remove", "list")
    }

    override fun onConsoleTabComplete(cs: ConsoleCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (args.size != 1) return emptyList()
        return getTabList(args[0], "add", "remove", "list")
    }
}